libraries/chain/hotstuff/finalizer.cpp

```diff
@@ -0,0 +1,302 @@
+   // from fork_db.root(), and specify weak.
+   auto bsp_last_qc = p->last_qc_block_num() ? get_block_by_height(p_branch, *p->last_qc_block_num()) : block_state_ptr{};
+
+   bool monotony_check = !fsi.last_vote || p->timestamp() > fsi.last_vote.timestamp;
+   // !fsi.last_vote means we have never voted on a proposal, so the protocol feature just activated and we can proceed
+
+   if (!fsi.lock.empty()) {
+      // Safety check : check if this proposal extends the proposal we're locked on
+      // --------------------------------------------------------------------------
+      if (extends(fork_db, p, fsi.lock.id))
+         safety_check = true;
+
+      // Liveness check : check if the height of this proposal's justification is higher
+      // than the height of the proposal I'm locked on.
+      // This allows restoration of liveness if a replica is locked on a stale proposal
+      // -------------------------------------------------------------------------------
+      if (bsp_last_qc && bsp_last_qc->timestamp() > fsi.lock.timestamp)
+         liveness_check = true;
+   } else {
+      // Safety and Liveness both fail if `fsi.lock` is empty. It should not happen.
+      // `fsi.lock` is initially set to `lib` when switching to IF or starting from a snapshot.
+      // -------------------------------------------------------------------------------------
+      liveness_check = false;
+      safety_check   = false;
+   }
+
+   dlog("liveness_check=${l}, safety_check=${s}, monotony_check=${m}",
+        ("l",liveness_check)("s",safety_check)("m",monotony_check));
+
+   // Figure out if we can vote and wether our vote will be strong or weak
+   // If we vote, update `fsi.last_vote` and also `fsi.lock` if we have a newer commit qc
+   // -----------------------------------------------------------------------------------
+   VoteDecision decision = VoteDecision::NoVote;
+
+   if (monotony_check && (liveness_check || safety_check)) {
+      auto [p_start, p_end] = std::make_pair(bsp_last_qc ? bsp_last_qc->timestamp() : p->timestamp(),
+                                             p->timestamp());
+
+      bool time_range_disjoint    = fsi.last_vote_range_start >= p_end || fsi.last_vote.timestamp <= p_start;
+
+      bool enough_for_strong_vote = time_range_disjoint || extends(fork_db, p, fsi.last_vote.id);
+
+      fsi.last_vote             = proposal_ref(p);          // v_height
+      fsi.last_vote_range_start = p_start;
+
+      if (chain.b1 && chain.b1->timestamp() > fsi.lock.timestamp)
+         fsi.lock = proposal_ref(chain.b1);                // commit phase on b1
+
+      decision = enough_for_strong_vote ? VoteDecision::StrongVote : VoteDecision::WeakVote;
+   } else {
+      dlog("bsp_last_qc=${bsp}, last_qc_block_num=${lqc}, fork_db root block_num=${f}",
+           ("bsp", !!bsp_last_qc)("lqc",!!p->last_qc_block_num())("f",fork_db.root()->block_num()));
+      if (p->last_qc_block_num())
```

Comment:
embed the if check inside the dlog

Reply:
Do you have a suggestion on how to do that cleanly?